pkg/pipeline/extract/timebased/timebased.go

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright (C) 2022 IBM, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package timebased
+
+import (
+	"container/list"
+
+	"github.com/netobserv/flowlogs-pipeline/pkg/api"
+	"github.com/netobserv/flowlogs-pipeline/pkg/config"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	OperationSum  = "sum"
+	OperationAvg  = "avg"
+	OperationMax  = "max"
+	OperationMin  = "min"
+	OperationLast = "last"
+	OperationDiff = "diff"
+)
+
+type FilterStruct struct {
+	Rule               api.TimebasedFilterRule
+	RecordKeyDataTable *RecordKeyTable
+	Results            filterOperationResults
+	Output             []filterOperationResult
```

Comment:
Must we keep both `Results` and `Output`?

Reply:
One is a map  and one is an array. All the computations are performed on the map, with all the results. However, we provide the user with an ordered array of only the topk/botk, so we have to produce the array. We keep it around until the next iteration for use in debugging and testing.

Reply:
So `Output` is written by `ComputeTopkBotk()` and then read by `CreateGenericMap()`.
https://github.com/netobserv/flowlogs-pipeline/blob/1bad271a90ba362bec08fee862c5095ea456e333/pkg/pipeline/extract/extract_timebased.go#L47-L49

I think I would write the code in way that `ComputeTopkBotk()` will return `Output` and pass it to `CreateGenericMap()`.
That would allow removing the `Output` field from `FilterStruct`.
But, I guess that's only a matter of style and I'm ok with leaving it as-is if I didn't convince you.